### ISSUE TITLE
Allow different error handling based on track type

### ISSUE THREAD

**Issue by amitks2015** [google/ExoPlayer#6640]
Yesterday our video streaming app ran into an issue where ExoPlayer was throwing ParserException. It was obviously due the corrupt subtitles(WebVTT) coming along with the HLS content. But it's not a good experience to prevent users from watching video because of corrupt subtitle data. 

My question is, in such cases, how can we let ExoPlayer continue playing the video and ignore the subtitle part? Is there a way we can disable the subtitle in this case?

**Comment by icbaker** [comment]
Could you give a bit more information? A stack trace showing where the exception is coming from would be helpful so we can best advise how to avoid/handle it.

A sample stream demonstrating the problem would also be very useful. If you're unable to share it here, you can email it to dev.exoplayer@gmail.com with this issue number in the subject line.

**Comment by amitks2015** [comment]
Thanks for your prompt response.
Here is the stack trace:
`<app package name> E/ExoPlayerImplInternal: Source error.
    com.google.android.exoplayer2.ParserException: Expected WEBVTT. Got Z���>gYx��(ٓW�I��T��z���x{�"��������inRzs��;
        at com.google.android.exoplayer2.text.webvtt.WebvttParserUtil.validateWebvttHeaderLine(WebvttParserUtil.java:44)
        at com.google.android.exoplayer2.source.hls.WebvttExtractor.processSample(WebvttExtractor.java:135)
        at com.google.android.exoplayer2.source.hls.WebvttExtractor.read(WebvttExtractor.java:127)
        at com.google.android.exoplayer2.source.hls.HlsMediaChunk.feedDataToExtractor(HlsMediaChunk.java:357)
        at com.google.android.exoplayer2.source.hls.HlsMediaChunk.loadMedia(HlsMediaChunk.java:325)
        at com.google.android.exoplayer2.source.hls.HlsMediaChunk.load(HlsMediaChunk.java:301)`

The stream was fixed by the backend team later so it's not available anymore.
But the problem was while playing the HLS(.m3u8 playlist) which has corrupt subtitle track.

**Comment by icbaker** [comment]
Thanks for the info!

Unfortunately I don't have a great answer for you.

You can modify the error handling behaviour by providing your own `LoadErrorHandlingPolicy` to the `HlsMediaSourceFactory`.

You can then disable the problematic track in your implementation of `getRetryDelayMsFor()`. You'd probably be best off extending `DefaultLoadErrorHandlingPolicy` and delegating all calls to that until it returns `C.TIME_UNSET` (i.e. this is a permanent failure) - then falling into your custom logic which would be something like:

```
if (isTextTrack) {
  handler.post(
      () -> trackSelector.setParameters(
          trackSelector
              .buildUponParameters()
              .setRendererDisabled(textRendererIndex, true)))
  // Make sure the error is never retried.
  return Long.MAX_VALUE;
} else {
  return C.TIME_UNSET
}
```

Note: You need to create the Handler on the main thread, before you construct the LoadErrorHandlingPolicy, the post the runnable back from inside, because the TrackSelector should only be interacted with from the main thread while the load error handling will be called on the loading thread.

You can get the `textRendererIndex` from `SimpleExoPlayer#getRendererType`.

Unfortunately, there's no sensible way currently to determine the `isTextTrack` bit. We have discussed adding more info to LoadErrorHandlingPolicy [internal: 144413506], which would allow you to make this determination reliably.

Right now, you could do it by snooping the exception you get passed to see if it comes from the `text` package or something, but this seems potentially brittle.

Assigning to Santiago as the owner of the internally tracked changes to LoadErrorHandlingPolicy.

**Comment by AquilesCanta** [comment]
Skipping load errors was considered and discarded back when we designed the load error handling policy. I'll try to find out why and update this post.

**Comment by icbaker** [comment]
The load might not be malformed for this to still be useful - imagine your subtitle file is 404ing but you'd still like to play the video & audio. You'd need to know the track type to make that decision.

**Comment by AquilesCanta** [comment]
By malformed loads I mean anything that is not successful. ParserException, non-HTTP-200, etc. Updated the comment to avoid confusion.

**Comment by amitks2015** [comment]
@icbaker Thanks for the workaround, I'll try and update here if it fixes the issue. But I would be nice if this issue can be contained within the library itself.

**Comment by VenomDino** [comment]
I was struggling with this issue and seached a lot but did not find any solution. Now I have fixed it by my self. I just added the below codes in `onPlayerError` callback and solved my problem.

`if (error.errorCode == PlaybackException.ERROR_CODE_PARSING_CONTAINER_MALFORMED && !hasRetried){

                    TrackSelectionParameters trackSelectionParameters = exoPlayer.getTrackSelectionParameters()
                            .buildUpon()
                            .setTrackTypeDisabled(C.TRACK_TYPE_TEXT, true)
                            .clearOverridesOfType(C.TRACK_TYPE_TEXT)
                            .build();
                    exoPlayer.setTrackSelectionParameters(trackSelectionParameters);
                    exoPlayer.prepare();

                    Toast.makeText(PlayerActivity.this, "Trying again because playback error.", Toast.LENGTH_SHORT).show();
                    hasRetried = true;
                }`

I just define a `boolean` variable called **hasRetried** to `false` and after retrying I simply changed the value to `false`. The reason behind declaring this variable is to prevent stucking in loop.

**Comment by tonihei** [comment]
I think the open feature request to add track-specific information to `LoadErrorHandlingPolicy` callbacks has been fixed since ExoPlayer 2.12.0. If there is any other missing information or open feature request, please file a new issue.
